### PR TITLE
PY-194: Add missing waitset status condition example for Python

### PR DESCRIPTION
**Pull request by luisrg-rti**

### CODE REVIEW

File: examples/connext_dds/waitset_status_cond/py/foo.py [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+# to this license. The software is provided "as is", with no warranty of any
+# type, including any warranty for fitness for any purpose. RTI is under no
+# obligation to maintain or support the software. RTI shall not be liable for
+# any incidental or consequential damages arising out of the use or inability
+# to use the software.
+
+import rti.idl as idl
+
+
+@idl.struct()
```
Comment (alexcamposruiz) [inline examples/connext_dds/waitset_status_cond/py/foo.py:14]:
```suggestion
@idl.struct
```

File: examples/connext_dds/waitset_status_cond/py/waitset_status_cond_publisher.py [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+def publisher_main(domain_id, sample_count):
+    # Create a DomainParticipant with default QoS
+    participant = dds.DomainParticipant(domain_id)
+
+    # Create a Topic and automatically register the type
+    topic = dds.Topic(participant, "Example waitset_status_cond", Foo)
+
+    # Create a DataWriter with default QoS (Publisher created in-line)
+    writer = dds.DataWriter(dds.Publisher(participant), topic)
```
Comment (alexcamposruiz) [inline examples/connext_dds/waitset_status_cond/py/waitset_status_cond_publisher.py:27]:
```suggestion
    writer = dds.DataWriter(topic)
```

File: examples/connext_dds/waitset_status_cond/py/waitset_status_cond_publisher.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+    # Instantiate a sample
+    sample = Foo()
+
+    # Write every second until the specified amount of samples is reached
+    count = 0
+    while (sample_count == 0) or (count < sample_count):
```
Comment (alexcamposruiz) [inline examples/connext_dds/waitset_status_cond/py/waitset_status_cond_publisher.py:34]: Let's also show the StatusCondition on the writer side

File: examples/connext_dds/waitset_status_cond/py/waitset_status_cond_subscriber.py [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+
+def subscriber_main(domain_id, sample_count):
+    # Create a DomainParticipant with default QoS
+    participant = dds.DomainParticipant(domain_id)
+
+    # Create a Topic and automatically register the type
+    topic = dds.Topic(participant, "Example waitset_status_cond", Foo)
+
+    # Create a DataReader with default QoS (Subscriber created in-line)
+    reader = dds.DataReader(dds.Subscriber(participant), topic)
```
Comment (alexcamposruiz) [inline examples/connext_dds/waitset_status_cond/py/waitset_status_cond_subscriber.py:26]:
```suggestion
    reader = dds.DataReader(topic)
```

File: examples/connext_dds/waitset_status_cond/py/waitset_status_cond_subscriber.py [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+
+        # Handle liveliness status changes
+        if dds.StatusMask.LIVELINESS_CHANGED in status_mask:
+            st = reader.liveliness_changed_status
+            print(f"Liveliness changed => Active writers = {st.alive_count}")
+
+        # Handle when there's new data available
+        if dds.StatusMask.DATA_AVAILABLE in status_mask:
+            nonlocal samples_read
+            for data, info in reader.take():
```
Comment (alexcamposruiz) [inline examples/connext_dds/waitset_status_cond/py/waitset_status_cond_subscriber.py:51]:
```suggestion
            for data in reader.take_data():
```

and remove he `if info.valid`.

File: examples/connext_dds/waitset_status_cond/py/waitset_status_cond_subscriber.py [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+
+from foo import Foo
+
+
+def subscriber_main(domain_id, sample_count):
+    # Create a DomainParticipant with default QoS
+    participant = dds.DomainParticipant(domain_id)
+
+    # Create a Topic and automatically register the type
+    topic = dds.Topic(participant, "Example waitset_status_cond", Foo)
```
Comment (alexcamposruiz) [inline examples/connext_dds/waitset_status_cond/py/waitset_status_cond_subscriber.py:23]: Use the same topic name as the other examples. May sure they interoperate.

```suggestion
    topic = dds.Topic(participant, "Example Foo", Foo)
```

File: examples/connext_dds/waitset_status_cond/py/USER_QOS_PROFILES.xml [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+                </subscription_name>
+            </datareader_qos>
+            <domain_participant_qos>
+                <!--
+                The participant name, if it is set, will be displayed in the
+                RTI tools, making it easier for you to tell one
+                application from another when you're debugging.
+                -->
+                <participant_name>
+                    <name>waitset_status_condParticipant</name>
```
Comment (alexcamposruiz) [inline examples/connext_dds/waitset_status_cond/py/USER_QOS_PROFILES.xml:46]: Use the same name pattern as in the other languages (e.g. ../c/USER_QOS_PROFILES.xml). Remove the role_name and the commented code below.

File: examples/connext_dds/waitset_status_cond/py/USER_QOS_PROFILES.xml [hunk 8]
```diff
@@ -9,59 +9,54 @@
...
-        <!-- QoS profile used to configure reliable communication between the DataWriter 
+    <qos_library name="waitset_status_cond_py_Library">
+        <!-- QoS profile used to configure reliable communication between the DataWriter
              and DataReader created in the example code.
-
              A QoS profile groups a set of related QoS.
         -->
-        <qos_profile name="waitset_status_cond_Profile" base_name="BuiltinQosLib::Generic.StrictReliable">
-            <!-- QoS used to configure the data writer created in the example code -->                
+        <qos_profile name="waitset_status_cond_py_Profile" base_name="BuiltinQosLibExp::Generic.StrictReliable" is_default_qos="true">
```
Comment (alexcamposruiz) [inline examples/connext_dds/waitset_status_cond/py/USER_QOS_PROFILES.xml:29]: The experiemental library is only maintained for backward compatibility, as far as I know. The reliability and history set below are redundant as well.
```suggestion
        <qos_profile name="waitset_status_cond_py_Profile" base_name="BuiltinQosLib::Generic.StrictReliable" is_default_qos="true">
```

File: examples/connext_dds/waitset_status_cond/py/waitset_status_cond_publisher.py [hunk 9]
```diff
@@ -21,25 +21,56 @@ def publisher_main(domain_id, sample_count):
...
+    send_samples = False
+
+    # Define a handler for the Status Condition
+    def status_handler(_):
+        nonlocal send_samples
+        status_mask = writer.status_changes
+        st = writer.publication_matched_status
+
+        if dds.StatusMask.PUBLICATION_MATCHED in status_mask:
+            if st.current_count > 0:
```
Comment (alexcamposruiz) [inline examples/connext_dds/waitset_status_cond/py/waitset_status_cond_publisher.py:45]: Print a message as you do in the subscriber

Comment (alexcamposruiz) [reply examples/connext_dds/waitset_status_cond/py/waitset_status_cond_publisher.py:45]: And I wouldn't control the publication of data based on matched subscribers. That normally goes against the publish/subscribe design pattern.

**Review by alexcamposruiz — CHANGES_REQUESTED** [review]

File: examples/connext_dds/waitset_status_cond/py/USER_QOS_PROFILES.xml (hunk 8 above)
Comment (luisrg-rti) [reply examples/connext_dds/waitset_status_cond/py/USER_QOS_PROFILES.xml:29]: I'll remove the `Exp`. I basically took it from the Modern C++ equivalent for simplicity:

https://github.com/rticommunity/rticonnextdds-examples/blob/master/examples/connext_dds/waitset_status_cond/c%2B%2B11/USER_QOS_PROFILES.xml

I suppose we can delete the `datawriter_qos`/`datareader_qos` part of the `.xml`, is that what you mean?

File: examples/connext_dds/waitset_status_cond/py/waitset_status_cond_publisher.py (hunk 9 above)
Comment (luisrg-rti) [reply examples/connext_dds/waitset_status_cond/py/waitset_status_cond_publisher.py:45]: I was going for some "utility", but I'll simply print the information and continue publishing the data as usual

File: examples/connext_dds/waitset_status_cond/py/waitset_status_cond_publisher.py [hunk 10]
```diff
@@ -32,20 +31,12 @@ def publisher_main(domain_id, sample_count):
...
-        nonlocal send_samples
-        status_mask = writer.status_changes
         st = writer.publication_matched_status
-
-        if dds.StatusMask.PUBLICATION_MATCHED in status_mask:
-            if st.current_count > 0:
-                send_samples = True
-            else:
-                send_samples = False
+        print(
```
Comment (alexcamposruiz) [inline examples/connext_dds/waitset_status_cond/py/waitset_status_cond_publisher.py:37]: Keep the if `if dds.StatusMask.PUBLICATION_MATCHED in status_mask:`

**Review by alexcamposruiz — CHANGES_REQUESTED** [review]

File: examples/connext_dds/waitset_status_cond/py/README.md [hunk 11]
```diff
@@ -0,0 +1,19 @@
+# Example Code: WaitSet with Status Condition
+
+If you haven't used the RTI Connext Python API before, first check the
+[Getting Started Guide](https://community.rti.com/static/documentation/connext-dds/7.4.0/doc/manuals/connext_dds_professional/getting_started_guide/index.html).
```
Comment (alexcamposruiz) [inline examples/connext_dds/waitset_status_cond/py/README.md:4]:
```suggestion
[Getting Started Guide](https://community.rti.com/static/documentation/connext-dds/current/doc/manuals/connext_dds_professional/getting_started_guide/index.html).
```

**Review by alexcamposruiz — CHANGES_REQUESTED** [review]

**Review by alexcamposruiz — APPROVED** [review]

**Pull request merged**